awx/main/tests/functional/api/test_workflow_node.py

```diff
@@ -132,11 +135,41 @@ def test_activity_stream_create_wf_approval(self, post, admin_user, workflow_job
 
         qs2 = ActivityStream.objects.filter(organization__isnull=True)
         assert qs2.count() == 5
-        assert qs2[0].operation == 'create'
-        assert qs2[1].operation == 'create'
-        assert qs2[2].operation == 'create'
-        assert qs2[3].operation == 'create'
-        assert qs2[4].operation == 'update'
+        assert list(qs2.values_list('operation', 'object1')) == [('create', 'user'),
+                                                                 ('create', 'workflow_job_template'),
+                                                                 ('create', 'workflow_job_template_node'),
+                                                                 ('create', 'workflow_approval_template'),
+                                                                 ('update', 'workflow_job_template_node'),
+                                                                 ]
+
+    @pytest.mark.django_db
+    def test_approval_node_actions(self, post, admin_user, job_template):
+        # This test ensures that a user (with permissions to do so) can approve/deny
+        # workflow approvals.  Also asserts that trying to approve/deny approvals
+        # that have already been dealt with will throw an error.
+        wfjt = WorkflowJobTemplate.objects.create(name='foobar')
+        node = wfjt.workflow_nodes.create(unified_job_template=job_template)
+        url = reverse('api:workflow_job_template_node_create_approval',
+                      kwargs={'pk': node.pk, 'version': 'v2'})
+        post(url, {'name': 'Approve/Deny Test', 'description': '', 'timeout': 0},
+             user=admin_user, expect=200)
+        post(reverse('api:workflow_job_template_launch', kwargs={'pk': wfjt.pk}),
+             user=admin_user, expect=201)
+        wf_job = WorkflowJob.objects.first()
+        TaskManager().schedule()
+        TaskManager().schedule()
+        wfj_node = wf_job.workflow_nodes.first()
+        approval = wfj_node.job
+        assert approval.name == 'Approve/Deny Test'
+        post(reverse('api:workflow_approval_approve', kwargs={'pk': approval.pk}),
+             user=admin_user, expect=204)
+        # Test that there is an activity stream entry that was created for the "approve" action.
+        qs = ActivityStream.objects.order_by('-timestamp').first()
+        assert qs.object1 == 'workflow_approval'
+        assert qs.changes == '{"status": ["pending", "successful"]}'
+        assert qs.operation == 'update'
```

Comment:
This looks great!

It would also be good to explicitly test that status e.g.,

```python
assert WorkflowApproval.objects.get(pk=approval.pk).status == 'successful'
```

Reply:
It would also be good to add another test just like this one that tests the `api:workflow_approval_deny` view (and similarly, verifies that the final status is `failed` and that the activity stream record has the `.changes` we anticipate.